README.md

```diff
@@ -1,26 +1,227 @@
+
+This example illustrates a simple service that implements a handler for a
+`Hello::echo` Thrift procedure.
+
+```thrift
+service Hello {
+    EchoResponse echo(1:EchoRequest echo)
+}
+
+struct EchoRequest {
+    1: required string message;
+    2: required i16 count;
+}
+
+struct EchoResponse {
+    1: required string message;
+    2: required i16 count;
+}
+```
+
+A go:generate directive informs `go generate` how to produce the Thrift models
+and YARPC bindings for the echo service.
+
+```go
+//go:generate thriftrw --plugin=yarpc echo.thrift
+```
+
+Setting up a YARPC dispatcher configures inbounds and outbounds for supported
+transport protocols and RPC types.
+This sets a service up to receive HTTP requests on port 8080 and send requests
+to itself.
+YARPC funnels requests from all inbound transports into its routing table, and
+organizes outbounds by name.
+
+```go
+httpTransport := http.NewTransport()
+dispatcher := yarpc.NewDispatcher(yarpc.Config{
+    Name: "hello",
+    Inbounds: yarpc.Inbounds{
+        httpTransport.NewInbound(":8080"),
+    },
+    Outbounds: yarpc.Outbounds{
+        "hello": {
+            Unary: httpTransport.NewSingleOutbound("http://127.0.0.1:8080"),
+        },
+    },
+})
+```
+
+The dispatcher governs the lifecycle of every inbound, outbound, and the
+singleton for each transport protocol.
+The singleton can manage the lifecycle of shared peers and connections.
+
+```go
+if err := dispatcher.Start(); err != nil {
+    log.Fatal(err)
+}
+defer dispatcher.Stop()
+```
+
+At the end of `main`, we block until we receive a signal to exit, then unravel
+anything deferred like `dispatcher.Stop()`, shutting down gracefully.
+
 ```
+signals := make(chan os.Signal, 1)
+signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+<-signals
+```
+
+### Handle
+
+To receive requests from any inbound, we register a handler object using the
+Thrift generated server.
+
+```go
+dispatcher.Register(helloserver.New(&helloHandler{}))
+```
+
+The handler must implement the Hello service from the Thrift IDL.
+The generated code requires a handler that implements `helloserver.Interface`.
+
+```go
+type helloHandler struct{}
+
+func (h helloHandler) Echo(ctx context.Context, e *echo.EchoRequest) (*echo.EchoResponse, error) {
+	return &echo.EchoResponse{Message: e.Message, Count: e.Count + 1}, nil
+}
+```
+
+### Call
+
+To send a request on an outbound, we construct a client using the corresponding
+named outbound from the dispatcher.
+The client will use that name for the outbound request service name
+
+```go
+client := helloclient.New(dispatcher.ClientConfig("hello"))
+```
+
+To call a remote procedure, we create a context with a one second deadline and
+call a method of the client.
+The client will use the dispatcher name for the caller name, Thrift for the
+encoding, and infer the procedure names from the `Echo` method (`Hello::echo`).
+
+```go
+ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+defer cancel()
```

Comment:
This reminds me: We should document somewhere that contexts must have a timeout or we will reject the request.